File: stock2.py
# // Time Complexity :O(n)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :no


# // Your code here along with comments explaining your approach

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        peak=prices[0]
        valley=prices[0]
        maxx=0
        i=0
        while(i<len(prices)-1):
            while(i<len(prices)-1 and prices[i]>=prices[i+1]):
                i+=1
            valley=prices[i]
            while(i<len(prices)-1 and prices[i]<=prices[i+1]):
                i+=1
            peak=prices[i]
           
            maxx=maxx+peak-valley
        return maxx
                
                
                
            

File: Stock2.java
// TC: O(n) where n is the length of the array.
// SC: O(1).
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Stock2{
    public int maxProfit(int[] prices) {
        if(prices.length == 0 || prices == null) return 0;
        int min, max, i = 0;
        int profit = 0;
        while(i <= prices.length - 2) {
            while(i < prices.length - 1 && prices[i] >= prices[i + 1]) {
                i++;
            }
            min = prices[i];
            while(i <= prices.length - 2 && prices[i] <= prices[i + 1]) {
                i++;
            }
            max = prices[i];
            profit = profit + max - min;
        }
        return profit;
    }
}